src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+      original_node = parser.parse
+
+      formatter.skip_space_or_newline
+      original_node.accept formatter
+
+      formatter.added_types? ? formatter.finish : nil
+    end
+
+    @_signatures : Hash(String, Signature)?
+
+    # Signatures represents a mapping of location => Signature for def at that location
+    def signatures : Hash(String, Signature)
+      @_signatures || raise "Signatures not properly initialized!"
+    end
+
+    # Given `accepted_defs` (location => (parsed) defs that match a def_locator), generated a new hash of
+    # location => (typed, multiple) def_instances that match a location.
+    #
+    # A given parsed def can have multiple def_instances, depending on how the method is called throughout
+    # the program, and the types of those calls.
+    private def accepted_def_instances(accepted_defs : Hash(String, Crystal::Def)) : Hash(String, Array(Crystal::Def))
+      ret = Hash(String, Array(Crystal::Def)).new do |h, k|
+        h[k] = [] of Crystal::Def
+      end
+
+      # First, check global definitions
+      program.def_instances.each do |_, def_instance|
+        next unless accepted_defs.keys.includes?(def_instance.location.to_s)
+
+        ret[def_instance.location.to_s] << def_instance
+      end
+
+      # Breadth first search time! This list will be a continuously populated queue of all of the types we need
+      # to scan, with newly discovered types added to the end of the queue from "parent" (namespace) types.
+      types = [] of Crystal::Type
+
+      program.types.each { |_, t| types << t }
+
+      overridden_method_locations = {} of String => String
+      while type = types.shift?
+        type.types?.try &.each { |_, t| types << t }
+        def_overrides_parent_def(type).each do |child_def_loc, ancestor_def_loc|
+          overridden_method_locations[child_def_loc] = ancestor_def_loc
+        end
+
+        # Check for class instance 'def's
+        if type.responds_to?(:def_instances)
+          type.def_instances.each do |_, def_instance|
+            next unless accepted_defs.keys.includes?(def_instance.location.to_s)
+
+            ret[def_instance.location.to_s] << def_instance
+          end
+        end
+
+        # Check for class 'self.def's
+        metaclass = type.metaclass
+        if metaclass.responds_to?(:def_instances)
+          metaclass.def_instances.each do |_, def_instance|
+            next unless accepted_defs.keys.includes?(def_instance.location.to_s)
+
+            ret[def_instance.location.to_s] << def_instance
+          end
+        end
+      end
+
+      # Now remove all overridden methods
+      overridden_method_locations.each do |child_loc, ancestor_loc|
+        if ret.delete(child_loc)
+          @warnings << "Not adding type restrictions to definition at #{child_loc} as it overrides definition #{ancestor_loc}"
+        end
+      end
+
+      ret
+    end
+
+    private def def_overrides_parent_def(type) : Hash(String, String)
```

Comment:
This method scans all `defs` in `type` and figures out the locations of all methods that override / are overridden between `type` and its ancestor's definitions. These methods should not have their types restricted in any additional way. This method is required as there's no syntatic way to indicate one method overrides another.